File: 164-maximum-gap/Solution.java
import java.util.Arrays;

class Solution164 {
    public static int maximumGap(int[] nums) {
        if (nums.length < 2)
            return 0;
        int maxNum = nums[0];
        for (int val : nums)
             maxNum = Math.max(val, maxNum);
        int maxDigitLen = 0;
        while (maxNum > 0){
             maxNum /= 10;
             maxDigitLen++;
        }
        int dev = 1;
        int[] counting = new int[10];
        for (int Digit = 0; Digit < maxDigitLen; Digit++) {
            Arrays.fill(counting, 0);
            for (int val : nums)
                counting[val / dev % 10]++;

            int preSum = 0;
            for (int index = 0; index < counting.length; index++) {
                int temp = counting[index];
                counting[index] = preSum;
                preSum += temp;
            }


            int[] result = new int[nums.length];
            for (int val : nums){
                int radix = val / dev % 10;
                result[counting[radix]] = val;
                counting[radix]++;
            }
            System.arraycopy(result, 0, nums, 0, nums.length);
            dev *= 10;
        }
        int maxGap = 0;
        for (int index = 1; index < nums.length; index++) {
            maxGap = Math.max(maxGap, (nums[index] - nums[index - 1]));
        }
         return maxGap;
    }
    static void main(String[] args) {
        int[] nums = {1,10000000};
        System.out.println("Ans = " + maximumGap(nums));
    }
}

File: 164-maximum-gap/solution.py
"""
參考代碼:
algo/RadixSort/radixSort_reverse.py
"""
from typing import List
class Solution:
    def maximumGap(self, nums: List[int]) -> int:
        max_num = nums[0]
        min_num = nums[0]
        for num in nums:
            if num > max_num:
                max_num = num
            elif num < min_num:
                min_num = num

        max_digit_len = 0
        while max_num > 0:
            max_num //= 10
            max_digit_len += 1
        dev = 1
        for digit in range(max_digit_len):
            counting = [0] * 10
            for num in nums:
                radix = num // dev % 10
                counting[radix] += 1
            print("Before prefix = ", counting)

            counting[0] -= 1
            for index in range(1, 10):
                counting[index] += counting[index - 1]
            print("After prefix = ", counting)

            result = [0] * len(nums)
            for index in range(len(nums) - 1, -1, -1):
                radix = nums[index] // dev % 10
                result[counting[radix]] = nums[index]
                counting[radix] -= 1
            dev *= 10
            nums[:len(nums)] = result[:len(result)]

        max_gap = 0
        for index in range(1, len(nums)):
            max_gap = max(max_gap, nums[index] - nums[index - 1])

        return max_gap
if __name__ == "__main__":
    arr = [3, 6, 9, 1]
    sol = Solution()
    print(sol.maximumGap(arr))